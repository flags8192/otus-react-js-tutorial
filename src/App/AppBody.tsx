import React, { FC } from 'react';
import styled from '@emotion/styled';
import GameField from '@src/modules/GameField';

const Root = styled.div`
  height: 80vh;
  position: relative;
`;

const AppBody: FC = () => {
  return (
    <Root role="gamePanel">
      <GameField />
    </Root>
  );
};

export default AppBody;
