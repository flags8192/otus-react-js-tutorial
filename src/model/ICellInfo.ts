enum ICellInfo {
  alive = 1,
  dead = 0,
}

export default ICellInfo;
