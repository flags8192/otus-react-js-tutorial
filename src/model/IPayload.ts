interface IPayload {
  gameFieldSize?: number;
  selected?: boolean;
  cellId?: number;
  gameLevel?: string;
}

export default IPayload;
