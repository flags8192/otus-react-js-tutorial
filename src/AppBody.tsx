import React, { Component } from 'react';
import styled from '@emotion/styled';
import { IUserList } from '@src/App';

const Root = styled.div`
  height: 70vh;
  align-items: center;
  justify-content: center;
  font-size: calc(10px + 2vmin);
  overflow-y: auto;
  margin: auto;
  width: 70%;

  p {
    text-align: center;
  }
`;

interface IAppBodyProps {
  data: IUserList;
}
export class AppBody extends Component<IAppBodyProps> {
  shouldComponentUpdate(nextProps: IAppBodyProps) {
    // Притянутая за уши оптимизация компонента
    return nextProps.data.length !== this.props.data.length;
  }

  render() {
    const { data } = this.props;

    return (
      <Root>
        {data.length ? (
          <ul role="dataList">
            {this.props.data.map((item) => (
              <li key={item.id} role="dataItem">
                {item.userId} {item.title}
              </li>
            ))}
          </ul>
        ) : (
          <p role="dataMessage">Нет данных</p>
        )}
      </Root>
    );
  }
}
