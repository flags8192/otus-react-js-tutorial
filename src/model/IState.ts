import ICellInfo from '@src/model/ICellInfo';
import { ISelectedCells } from '@src/modules/Cell/models';

export interface IState {
  gameFieldSize: number;
  gameFieldPercentFilled: number;
  gameFieldData: Array<ICellInfo>;
  gameLevel: string;
  selectedCells: ISelectedCells;
}
