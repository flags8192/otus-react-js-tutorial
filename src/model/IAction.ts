import IPayload from './IPayload';

export interface IAction {
  type: string;
  payload?: IPayload;
}
