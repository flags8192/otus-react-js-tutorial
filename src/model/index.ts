import { IState } from '@src/model/IState';
import { IAction } from '@src/model/IAction';

export { IAction, IState };
