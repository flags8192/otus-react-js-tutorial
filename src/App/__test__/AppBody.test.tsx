import React from 'react';
import { render, screen } from '@testing-library/react';

import AppBody from '@src/App/AppBody';
import Store from '@src/store';
import { IAction } from '@src/model';

let dispatch: (action: IAction) => void;

describe('AppBody', () => {
  beforeEach(() => {
    dispatch = jest.fn();
  });

  it('Render <AppBody> without data', () => {
    const state = {
      gameLevel: '1',
      gameFieldSize: 3,
      gameFieldPercentFilled: 10,
      gameFieldData: [1, 0, 0, 0, 0, 0, 0, 0, 0],
      selectedCells: {},
    };
    const { asFragment } = render(
      <Store.Provider value={{ dispatch, state }}>
        <AppBody />
      </Store.Provider>
    );
    expect(asFragment()).toMatchSnapshot();
    const dataMessage = screen.getByRole(/gamePanel/gi);
    expect(dataMessage).toBeInTheDocument();
  });

  it('Render <AppBody> with data', () => {
    const state = {
      gameLevel: '1',
      gameFieldSize: 4,
      gameFieldPercentFilled: 10,
      gameFieldData: [1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
      selectedCells: {},
    };
    const { asFragment } = render(
      <Store.Provider value={{ dispatch, state }}>
        <AppBody />
      </Store.Provider>
    );
    expect(asFragment()).toMatchSnapshot();
    const dataList = screen.getByRole(/gamePanel/gi);
    expect(dataList).toBeInTheDocument();
    const dataItems = screen.getAllByRole(/cellContainer/gi);
    expect(dataItems.length).toBe(9);
  });
});
