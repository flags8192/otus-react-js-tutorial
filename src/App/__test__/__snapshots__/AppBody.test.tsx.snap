// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppBody Render <AppBody> with data 1`] = `
<DocumentFragment>
  <div
    class="css-xt4orx"
    role="gamePanel"
  >
    <section
      class="css-hsjawn"
      role="cellsGrid"
    >
      <div
        class="css-es8rv7"
        role="cellContainer-0"
      >
        <div
          class="css-w6oewy"
          role="cellContent-0"
        />
      </div>
      <div
        class="css-2r18jt"
        role="cellContainer-1"
      >
        <div
          class="css-w6oewy"
          role="cellContent-1"
        />
      </div>
      <div
        class="css-2r18jt"
        role="cellContainer-2"
      />
      <div
        class="css-1jbr29x"
        role="cellContainer-3"
      />
      <div
        class="css-es8rv7"
        role="cellContainer-4"
      />
      <div
        class="css-2r18jt"
        role="cellContainer-5"
      />
      <div
        class="css-2r18jt"
        role="cellContainer-6"
      />
      <div
        class="css-1jbr29x"
        role="cellContainer-7"
      />
      <div
        class="css-es8rv7"
        role="cellContainer-8"
      />
      <div
        class="css-2r18jt"
        role="cellContainer-9"
      />
      <div
        class="css-2r18jt"
        role="cellContainer-10"
      />
      <div
        class="css-1jbr29x"
        role="cellContainer-11"
      />
      <div
        class="css-es8rv7"
        role="cellContainer-12"
      />
      <div
        class="css-2r18jt"
        role="cellContainer-13"
      />
      <div
        class="css-2r18jt"
        role="cellContainer-14"
      />
      <div
        class="css-1jbr29x"
        role="cellContainer-15"
      />
      <div
        style="clear: both;"
      />
    </section>
  </div>
</DocumentFragment>
`;

exports[`AppBody Render <AppBody> without data 1`] = `
<DocumentFragment>
  <div
    class="css-xt4orx"
    role="gamePanel"
  >
    <section
      class="css-hsjawn"
      role="cellsGrid"
    >
      <div
        class="css-es8rv7"
        role="cellContainer-0"
      >
        <div
          class="css-w6oewy"
          role="cellContent-0"
        />
      </div>
      <div
        class="css-2r18jt"
        role="cellContainer-1"
      />
      <div
        class="css-1jbr29x"
        role="cellContainer-2"
      />
      <div
        class="css-es8rv7"
        role="cellContainer-3"
      />
      <div
        class="css-2r18jt"
        role="cellContainer-4"
      />
      <div
        class="css-1jbr29x"
        role="cellContainer-5"
      />
      <div
        class="css-es8rv7"
        role="cellContainer-6"
      />
      <div
        class="css-2r18jt"
        role="cellContainer-7"
      />
      <div
        class="css-1jbr29x"
        role="cellContainer-8"
      />
      <div
        style="clear: both;"
      />
    </section>
  </div>
</DocumentFragment>
`;
